File: frontend/iosApp/Domains/Sources/Domains/SyncOperation.swift
import Foundation

public struct SyncOperation: Sendable, Identifiable, Hashable {
    public let id: UUID
    public let entityType: EntityType
    public let operationType: OperationType
    public let localId: UUID
    public let createdAt: Date
    public var status: SyncOperationStatus
    public var errorMessage: String?

    public init(
        id: UUID,
        entityType: EntityType,
        operationType: OperationType,
        localId: UUID,
        createdAt: Date,
        status: SyncOperationStatus,
        errorMessage: String? = nil
    ) {
        self.id = id
        self.entityType = entityType
        self.operationType = operationType
        self.localId = localId
        self.createdAt = createdAt
        self.status = status
        self.errorMessage = errorMessage
    }
}

public enum EntityType: String, Codable, Sendable {
    case album
    case memory
    case user
}

public enum OperationType: String, Codable, Sendable {
    case create
    case update
}

public enum SyncOperationStatus: String, Codable, Sendable {
    case pending
    case inProgress
    case failed
}

File: frontend/shared/src/commonMain/kotlin/com/example/memoriesapp/domain/SyncOperation.kt
package com.example.memoriesapp.domain

import com.example.memoriesapp.core.LocalId
import com.example.memoriesapp.core.Timestamp
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

data class SyncOperation(
    val id: LocalId,
    val entityType: EntityType,
    val operationType: OperationType,
    val localId: LocalId,
    val createdAt: Timestamp,
    val status: SyncOperationStatus,
    val errorMessage: String? = null
)

@Serializable
enum class EntityType {
    @SerialName("album")
    ALBUM,

    @SerialName("memory")
    MEMORY,

    @SerialName("user")
    USER
}

@Serializable
enum class OperationType {
    @SerialName("create")
    CREATE,

    @SerialName("update")
    UPDATE
}

@Serializable
enum class SyncOperationStatus {
    @SerialName("pending")
    PENDING,

    @SerialName("inProgress")
    IN_PROGRESS,

    @SerialName("failed")
    FAILED
}
